### PR TITLE
token invalidation

### LINKED ISSUE #3: Invalidate authentication token
It would be nice to be able to invalidate authentication tokens on request.

### PR DESCRIPTION
**Pull request by albalitz**
This introduces functionality allowing for manually invalidating auth tokens. See #3 

When generating a new token, it is now saved to the database. Then, when the user uses a token to authenticate, the database is checked for it.
If it exists and is valid, everything works as before. If it exists in the database, but is expired, it is deleted from the database (no need to keep it there).

Additionally, this introduces two new routes: /token/invalidate and /token/invalidate/all
See [README.md@ec8cd12](https://github.com/albalitz/save-api/blob/ec8cd1215cd446a5463e58560fc5e0bde3942355/README.md) for usage details.

@mspl13, since you opened #3 and asked for this, I think it'd be good if you checked this before I merge it, to see if this works as expected. ;)

P.S.: When merged, this closes #3

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -77,6 +77,30 @@ When using this route with the username/password combination, it returns a json
 Using an invalid token fails the authentication and returns a 401 unauthorized.
 
 
+### invalidate auth tokens
+Invalidating auth tokens can be done one by one which requires an auth token to be used,
+or all tokens from the authenticating user at once.
+
+#### invalidate the current token
+```
+curl -X DELETE -u <token>:"" http(s)://<your_domain>/save/api/token/invalidate
```
Comment (mspl13) [inline README.md]: I don't know if this is a problem but wouldn't this method allow anyone to invalidate any token?

Comment (mspl13) [inline README.md]: I don't know if this is a problem but wouldn't this method allow anyone to invalidate any token?

File: src/models.py [hunk 2]
```diff
@@ -89,12 +106,23 @@ def generate_auth_token(self, expiration = 600):
...
+            return None  # token not in the database => invalidated?
+
         s = Serializer(SECRET_KEY)
         try:
             data = s.loads(token)
             app.logger.info("Token verified.")
         except SignatureExpired as e:
             app.logger.warning(e)
+
+            # no need to keep expired tokens in the database
```
Comment (mspl13) [inline src/models.py]: What happens in the following case: The user never logs out manually and starts collecting (in)valid tokens in the database? Wouldn't it be wise to also delete invalid tokens on every x-th login or so?

File: README.md (hunk 1 above)
Comment (albalitz) [inline README.md]: When that someone has your token, yes. The same is true for every other route when using the token authentication.
For that case, the invalidate all route works with username/password authentication, so even when that happens, you can always use your username/password to invalidate all your tokens, rendering the token someone gained access to useless.

File: src/models.py (hunk 2 above)
Comment (albalitz) [inline src/models.py]: When trying an invalid token to do anything, it is removed from the database.
If someone just generates a lot of tokens and never uses them, that would indeed fill the database...

> Wouldn't it be wise to also delete invalid tokens on every x-th login or so?

Something similar would probably be good, yes. I could simply check every token from the authenticating user and remove the invalid ones - that wouldn't require counting logins etc.

Comment (albalitz) [inline src/models.py]: Added with [cbb6e85](https://github.com/albalitz/save-api/pull/4/commits/cbb6e85a279d87ed0b168ae59a57a18dff43dd28).

File: README.md (hunk 1 above)
Comment (mspl13) [inline README.md]: Fair enough.

**Pull request merged**